app/src/migrations/20200511000001-7-audit-trigger.js

```diff
@@ -0,0 +1,137 @@
+module.exports = {
+  up: (queryInterface) => {
+    return queryInterface.sequelize.transaction((t) => {
+      return queryInterface.sequelize.query('CREATE TABLE logged_actions (\n' +
+        '    schema_name text NOT NULL,\n' +
+        '    TABLE_NAME text NOT NULL,\n' +
+        '    user_name text,\n' +
+        '    action_tstamp TIMESTAMP WITH TIME zone NOT NULL DEFAULT CURRENT_TIMESTAMP,\n' +
+        '    action TEXT NOT NULL CHECK (action IN (\'I\',\'D\',\'U\')),\n' +
+        '    original_data text,\n' +
+        '    new_data text,\n' +
+        '    query text\n' +
+        ') WITH (fillfactor=100)',
+      {
+        transaction: t
+      }).then(() => {
+        return queryInterface.sequelize.query('REVOKE ALL ON logged_actions FROM public;',
+          {
+            transaction: t
+          });
+      }).then(() => {
+        return queryInterface.sequelize.query('GRANT SELECT ON logged_actions TO public;',
+          {
+            transaction: t
+          });
+      }).then(() => {
+        return queryInterface.sequelize.query('CREATE INDEX logged_actions_schema_table_idx ON logged_actions(((schema_name||\'.\'||TABLE_NAME)::TEXT));',
+          {
+            transaction: t
+          });
+      }).then(() => {
+        return queryInterface.sequelize.query('CREATE INDEX logged_actions_action_tstamp_idx ON logged_actions(action_tstamp);',
+          {
+            transaction: t
+          });
+      }).then(() => {
+        return queryInterface.sequelize.query('CREATE INDEX logged_actions_action_idx ON logged_actions(action);',
+          {
+            transaction: t
+          });
+      }).then(() => {
+        return queryInterface.sequelize.query('CREATE OR REPLACE FUNCTION if_modified_func() RETURNS TRIGGER AS $body$\n' +
+          'DECLARE\n' +
+          '    v_old_data TEXT;\n' +
+          '    v_new_data TEXT;\n' +
+          'BEGIN\n' +
+          '    /* This dance with casting the NEW and OLD values to a ROW is not necessary in pg 9.0+ */\n' +
```

Comment:
> not necessary in pg 9.0+ 
🤔

Reply:
I don't know PG well enough to sort through what is and isn't necessary and since it works as advertised, I just left it alone.  Better than introducing some bug or weird behaviour.  If there is time, maybe someone can figure out what is extraneous.

Reply:
I'll do a quick check and see if it still works without the cast.

Reply:
For sure, if it works it works which is great. Just noting we're on 9+

Reply:
I took out the ROW(NEW.*) and it didn't work. Will have to look at other examples of triggers to figure it out.

Reply:
Tried a few things, couldn't figure it out quickly. :(. Leaving as is.